### DISCUSSION TITLE
Reading the ~/.ssh/config file

### DISCUSSION THREAD

**Discussion by andylaw** [Eugeny/tabby#10017] [Q&A]
I've got a lot of SSH configuration set up in my OpenSSH config file. Tabby makes a stab at reading it, and kind-of does well enough but the file parsing code is naive when it comes to inheritance of configuration. I started to look at re-writing the function, but then I began to wonder if it was actually necessary. The code stores some features of the settings into the SSHProfile structures but does it actually use them or does it rely on OpenSSH to have done its _own_ parsing for the actual functionality.

If the answer is "yes, Tabby does use all the SSHProfile attributes and makes its own connections" then I'll try to take a better look at improving the code and implementing the inheritance rules. If not, then I'll look at how the code might be modified to be smarter about parsing the host names.

As an example of what is giving me grief at the moment, if I have two hosts with mostly shared details, but just a single distinguishing feature between them, then I have the following in my config file...

```
# Shared stuff here
Host  conn_1 conn_2
    ...shared details...

Host  conn_1
    ...conn_1 specifics...

Host  conn_2
    ...conn_2 specifics...

```

That _should_ give me two profiles in the "Imported from SSH config" group, but it gives me 3 i.e. `conn_1`, `conn_2` and `conn_1 conn_2`, the last of which is bogus

**Comment by Eugeny** [comment]
You're correct! Currently the OpenSSH config parser is somewhat naive and only covers basic syntax. 
The file is parsed into `SSHProfile`s which are then provided as built-in profiles that you connect to, Tabby doesn't interact with OpenSSH directly at any point

**Reply by andylaw** [reply]
So in my example above, the `conn_1` profile wouldn't work properly because it doesn't inherit the shared parts? I suspected that Tabby was just calling `ssh conn_1` under the hood, leaving the heavy lifting to SSH but now I realise that's not the case. Is there an open issue relating to this that I can look to hop onto, assuming I can find any spare time?

**Reply by Eugeny** [reply]
Yes! I think this is the closest one: https://github.com/Eugeny/tabby/issues/10012

If you decide to work on this, the place to look at is the `OpenSSHImporter` class in https://github.com/Eugeny/tabby/blob/master/tabby-electron/src/sshImporters.ts#L15

**Reply by andylaw** [reply]
Working on this now. I confess that Typescript is not my first language, so apologies for daft questions, but why is `sshimporters.ts` in the `tabby-electron` package and not the `tabby-ssh` package?

I think I have a mostly functional function that reads in the SSH config file and converts all the 'complete' (no wildcards in names, includes at least a Hostname) Hosts into `SSHProfile` objects. It relies on an additional third party library to do the heavy lifting. As I said, Typescript is not my tool of choice for day-to-day work, so I'm not sure which `package.json` file I should add that to (the one inside `tabby-electron`, or the top-level one), nor whether it should in `dependencies` (where I _think_ it should be), or `dev-dependencies` which is where I can see some libraries that I suspect are required in the final packaged app.

I'm hoping to have a pull-request available for initial review and feedback sometime through the week, but paid work may get the way of that

**Reply by Eugeny** [reply]
The reason for separation is because it has a NodeJS dependency (`fs`) and core packages are nodejs-free.

New dependencies go into their respective plugin's `package.json`, into `devDependencies` if they are going to be bundled by webpack (default) or `dependencies` if they are a webpack external module.

**Comment by andylaw** [comment]
@Eugeny - I've now got code that's successfully reading in the SSHConfig file and converting it into SSHProfile objects. It returns a list of SSHProfile objects having parsed all the wildcard and multiple entries for a single 'Host' and the list obtained contains a single, complete entry for each Host that has at least a Hostname specified.

The problem that I've now run into is that Hosts that define JumpHosts fail because Tabby doesn't find the JumpHost in its config. I _suspect_ that this may be because it only searches in its own config, rather than the externally-defined SSHConfig because I can see the jump host in the list of config-derived profiles and it connects as I would expect if I use it explicitly. Is that guess likely to be on the money?

I'm starting at Line 79 in `sshTab.component.ts` under tabby-sh. Any thoughts before I dig too much deeper?

**Reply by andylaw** [reply]
OK. I can now see that I'm storing the JumpHost ID as the Host name from the ssh config file, but the search for the matching profile uses the ID which prefixes the JumpHost Host name with `openssh-config:`. Again, I'm guessing that the rationale is to allow SSH config to have a 'profile' with the same name as one defined to use telnet or some other option. However, I think that there's a key argument to be made that the SSH-config and the Tabby-defined ssh profiles do need to overlap so I should be able to reference one from the other

**Reply by Eugeny** [reply]
Correct - you can still use an OpenSSH profile as a jumphost for a Tabby native profile, but not the other way around because imported profiles are read-only. So when you import a profile, you need to match the `jumpHost` ID to what OpenSSHImporter is generating as an ID.

**Reply by andylaw** [reply]
Right. That makes sense, and my initial tests storing the JumpHost in that way seem to work.

**Reply by peix187** [reply]
Hi guys,

Thanks for your work. I download the last build from https://nightly.link/Eugeny/tabby/workflows/build/master and i have the same problem #10012

**Reply by andylaw** [reply]
@peix187 - are you saying that the problem is not fixed in the last nightly build? If so, I doubt that the patch has landed in the nightly just yet since the pull request only got accepted at lunchtime.

**Reply by peix187** [reply]
@andylaw The nightly version is build in few minutes after merge but it's right your patch seem not be build https://github.com/Eugeny/tabby/actions/runs/11702834479/job/32591956008
Windows packages are failed.
@Eugeny can you check this, please?

**Reply by andylaw** [reply]
@peix187 - I wasn't aware that the nightly gets built on a commit.

On a separate issue, the code (when it can be persuaded to compile properly) currently treats SSH Config settings and the profiles created within Tabby as separate things. Are there circumstances where you would ever want the SSH Config to provide a platform on which you built Tabby profiles. For example, defining a set of wildcard configurations in SSH Config and then using those to run tabby profiles with a minimal set of additional data added in?

I don't have that need myself (just now), but I can readily imagine it being a requirement for _someone_.

**Reply by peix187** [reply]
@andylaw, yes i use wildcard in my configuration. Look this simple configuration and the result in Tabby profiles.

```
Host srv-a
    Hostname 172.22.0.1

Host srv-*
    User admin
    IdentitiesOnly yes
    IdentityFile ~/.ssh/mysshkey

Host *
    NumberOfPasswordPrompts 5
```

![image](https://github.com/user-attachments/assets/aae0741d-f399-4f9e-92f1-e209a191a6ba)

For me Tabby should display only **one** profile (with a _Hostname_ parameter) but apply wildcard blocks for all correspondent "normal" block.

**Reply by Eugeny** [reply]
I'm currently working on getting validated for a new code signing certificate, until then windows builds have to be built manually

**Reply by andylaw** [reply]
@peix187 - your scenario works as both you and I would expect in my build. Hopefully once the Windows build is back up and running then you'll be able to check that. However, I was asking about a potential situation where your wildcards were defined in SSH Config, but your specific Hosts were created by adding a Tabby Profile using the "New" button in the Profiles and Connections tab. Would you expect/want a hostname entered in that profile to be processed through the SSH Config settings to automatically fill in details?

**Reply by peix187** [reply]
@andylaw - Thank for your propose but i use only the SSH Config file for define hosts. If i have time this night i wild try to build the windows package for testing your patch.

**Reply by peix187** [reply]
Hi guys, sorry for the late feedback. I have build app with the last commit (ac6f60f) and when i launch tabby.exe (portable) nothing appear (windows or taskbar). The process is running because i see it in the task manager. Launch as administrator is the same. No error on cli with --debug or --inspect. An idea?

**Reply by peix187** [reply]
Hi @andylaw, i hope you are well. With build 1.0.216 my profiles list is empty (no **Imported from .ssh/config** group). I use the config file like my comment [here](https://github.com/Eugeny/tabby/discussions/10017#discussioncomment-11171435)

**Reply by peix187** [reply]
1.0.217 seems to work correctly.

**Comment by andylaw** [comment]
Sorry, I am of no help here. I’m a Mac user.

From: peix187 ***@***.***>
Date: Saturday, 23 November 2024 at 01:19
To: Eugeny/tabby ***@***.***>
Cc: Andy Law ***@***.***>, Mention ***@***.***>
Subject: Re: [Eugeny/tabby] Reading the ~/.ssh/config file (Discussion #10017)
This email was sent to you by someone outside the University.
You should only click on links or attachments if you are certain that the email is genuine and the content is safe.

Hi guys, sorry for the late feedback. I have build app with the last commit (ac6f60f<https://github.com/Eugeny/tabby/commit/ac6f60f1ae9ca754b22f15e06cf55a7574b1f9d5>) and when i launch tabby.exe (portable) nothing appear (windows or taskbar). The process is running because i see it in the task manager. Launch as administrator is the same. No error on cli with --debug or --inspect. An idea?

—
Reply to this email directly, view it on GitHub<https://github.com/Eugeny/tabby/discussions/10017#discussioncomment-11354472>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAKKLUZZWQG6JDMRK5FKPJD2B7JY7AVCNFSM6AAAAABQNL3EH6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMZVGQ2DOMQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

The University of Edinburgh is a charitable body, registered in Scotland, with registration number SC005336. Is e buidheann carthannais a th’ ann an Oilthigh Dhùn Èideann, clàraichte an Alba, àireamh clàraidh SC005336.

**Reply by peix187** [reply]
@andylaw - I have unpacked the asar file and i don't see ssh-config folder in node_modules. Why you put ssh-config package in devDependencies?

**Reply by andylaw** [reply]
@peix187 - because that was my understanding of what @Eugeny told me earlier in this discussion

**Reply by Eugeny** [reply]
@peix187 it's compiled into the webpack bundle and is not in the release node_modules. The .ssh/config importer works for me in the nightly on a mac. Any devtools console errors?
